text/0000-cargo-schema-version.md

```diff
@@ -0,0 +1,255 @@
+named `build.rs` by convention, whereas old Cargo required specifying `build =
+"build.rs"` explicitly.  If old Cargo attempted to use new crates relying on
+these features, it would fail at build time with difficult-to-diagnose errors.
+
+In order to introduce further changes to Cargo in the future, we propose
+introducing a Cargo interface/schema version, and a means for both `Cargo.toml`
+files and crate registry indexes (such as the one managed by crates.io) to
+incorporate that version.  Older versions of Cargo can then detect that version
+and skip crates requiring new Cargo at dependency resolution time, or emit
+meaningful errors at build time.
+
+Schema versioning represents a dependency for multiple subsequent Cargo
+features, including:
+
+- stdlib-aware Cargo (RFC #1133)
+- Improvements to cross-compilation support
+- Dependencies on the version or features of the Rust language
+- Dependencies on external system libraries (Issue rust-lang/cargo/#3816)
+- Dependencies on arbitrary tools to be used at compile-time (such as bindgen,
+  gcc, or even rustc itself)
+- Non-cargo tools parsing `Cargo.toml` and/or the index (such as to generate
+  Linux distribution packages, or to integrate into an [external build
+  system](https://github.com/rust-lang/rust-roadmap/issues/12))
+
+While some of these changes could potentially be accomplished by adding new
+fields to `Cargo.toml` that old versions of Cargo would ignore, this presents a
+significant downside in usability and functionality: A crate author using these
+features *cannot* ensure that a user with an old version of Cargo will build
+the code in the manner intended.  Instead, the crate author must either attempt
+to cope with the failure modes of all past versions of Cargo (and the various
+fields they ignore), or attempt some kind of ad-hoc mechanism for enforcing
+failure in such cases.  However, any such mechanism cannot take effect any
+earlier than at build time, and thus makes it impossible for Cargo to fall back
+to a different version of that crate which *would* have worked.  Such a
+mechanism would also likely express a stricter version requirement on the Cargo
+tool than necessary, and break compatibility with non-Cargo tools parsing
+`Cargo.toml` files.  The schema version introduced in this RFC allows Cargo to
+take that version into account in dependency resolution.
+
+This change will also allow for future experiments with new "unstable" Cargo
+features not yet ready for use in stable crates, without allowing those
+features to leak into stable crates or have their behavior set in stone.  Any
+future RFC introducing such behavior should consider whether crates.io should
+prohibit the use of such features entirely, or keep them separated and tagged
+appropriately in indexes.
+
+# Detailed design
+[design]: #detailed-design
+
+Introduce a new Cargo schema version, initially `1.0.0`.  Cargo will increase
```

Comment:
I'm not sure how much we gain if anything from having a full semver-style version here. docker-compose just uses a single integer, for example. (I haven't read the RFC in detail yet, so this may have been covered elsewhere already 😄 )